include/deal.II/lac/matrix_scaling.h

```diff
@@ -0,0 +1,410 @@
+       */
+      unsigned int start_inf_norm_steps;
+
+      /**
+       * Number of intermediate l1 norm scaling steps.
+       */
+      unsigned int l1_norm_steps;
+
+      /**
+       * Number of final linfty norm scaling steps.
+       */
+      unsigned int end_inf_norm_steps;
+    };
+
+    /**
+     * Construct a new AdditionalData object with the given parameters.
+     */
+    explicit AdditionalData(
+      const double           scaling_tolerance = 1e-5,
+      const ScalingAlgorithm alg =
+        ScalingAlgorithm::l1_linf_symmetry_preserving,
+      const SKParameters     sk_params     = SKParameters(),
+      const l1linfParameters l1linf_params = l1linfParameters());
+
+
+    /**
+     * Tolerance for the scaling algorithms. Convergence is achieved if the
+     * difference in norm between a vector of all ones and the row and column
+     * norms vectors is below this tolerance.
+     * Convergence is not needed to obtain a good scaling and reaching it might
+     * require a lot of iterations and thus time. Convergence is not guaranteed
+     * for all kind of matrices. For more details check out the articles cited
+     * in the class description.
+     */
+    double scaling_tolerance;
+
+    /**
+     * Scaling algorithm to use.
+     */
+    ScalingAlgorithm algorithm;
+
+    /**
+     * Parameters for the Sinkhorn-Knopp algorithm.
+     */
+    SKParameters sinkhorn_knopp_parameters;
+
+    /**
+     * Parameters for the symmetry preserving scaling algorithm.
+     */
+    l1linfParameters l1linf_parameters;
+  };
+
+  /**
+   * Constructor. The constructor takes a <tt>AdditionalData</tt> object that
+   * contains the parameters for the scaling algorithms and the algorithm
+   * selected.
+   */
+  explicit MatrixScaling(const AdditionalData &control = AdditionalData());
+
+  /**
+   * Destructor.
+   */
+  ~MatrixScaling() = default;
+
+  /**
+   * Scale the input <tt>matrix</tt> in place according to the selected
+   * algorithm in <tt>AdditionalData</tt>.
+   *
+   * The function returns whether the scaling algorithm has converged or not.
+   * Convergence is not necessary to have a good scaling but it can be helpful
+   * to know if it was achieved.
+   */
+  template <class Matrix>
+  bool
+  scale_matrix(Matrix &matrix);
+
+  /**
+   * Scale the system <tt>matrix</tt> in place according to the selected
+   * algorithm in <tt>AdditionalData</tt> and scale in place the right hand side
+   * vector <tt>rhs</tt> according to the row scaling. The solution of the
+   * scaled system can be scaled back to the original system by calling
+   * <tt>scale_system_solution()</tt>.
+   *
+   * The function returns whether the scaling algorithm has converged or not.
+   * Convergence is not necessary to have a good scaling but it can be helpful
+   * to know if it was achieved.
+   */
+  template <class Matrix, class VectorType>
+  bool
+  scale_linear_system(Matrix &matrix, VectorType &rhs);
```

Comment:
As mentioned in https://github.com/dealii/dealii/pull/18880#discussion_r2449797913, I think these two functions should be renamed. A long name that expresses what you do would be `determine_scaling_and_scale_{matrix,linear_system}`, but if you come up with a different name (perhaps as suggested in the linked-to comment), that's fine with me too.